### PR TITLE
Infinispan - Support caching annotations

### PR DESCRIPTION
**Pull request by karesti**
Caching annotations @CacheInvalidate @CacheInvalidateAll and @CacheResult

https://issues.redhat.com/browse/ISPN-13062

### CODE REVIEW

**Comment by gsmet** [comment]
@karesti thanks but could you split that in two different commits? If we have to backport an Infinispan upgrade, I would like to be able to do it easily. Thanks.

**Comment by karesti** [comment]
https://github.com/quarkusio/quarkus/pull/25306

**Review by gwenneg — COMMENTED** [review]
Here's a review focused on the code inspired/copied from `quarkus-cache`. Please double check the Infinispan-related code as it's not something I know in details :smile:

File: docs/src/main/asciidoc/infinispan-client.adoc
Comment (karesti) [reply docs/src/main/asciidoc/infinispan-client.adoc:464]: In the Spring cache integration we use this kind of mechanisms as well. I did not mention it in the documentation I wanted a review from @wburns on this part before that

File: docs/src/main/asciidoc/infinispan-client.adoc
Comment (karesti) [reply docs/src/main/asciidoc/infinispan-client.adoc:460]: use the method argument as the cache key

File: extensions/infinispan-client/deployment/src/main/java/io/quarkus/infinispan/client/deployment/CacheNamesBuildItem.java
Comment (karesti) [reply extensions/infinispan-client/deployment/src/main/java/io/quarkus/infinispan/client/deployment/CacheNamesBuildItem.java:10]: I have some stuff I forgot to remove thanks ! I considered this for a moment and then realised we did not yet need that @gwenneg

File: extensions/infinispan-client/deployment/src/main/java/io/quarkus/infinispan/client/deployment/InfinispanClientProcessor.java
Comment (karesti) [reply extensions/infinispan-client/deployment/src/main/java/io/quarkus/infinispan/client/deployment/InfinispanClientProcessor.java:93]: yes. if I don't add them the interceptors are not discoverable and instanced. I don't use the beans.xml file in the extension, since some beans are optional (health checks for example) and needed only when we use the health check extension and the infinispan client extension at the same time.

File: extensions/infinispan-client/deployment/src/main/java/io/quarkus/infinispan/client/deployment/InfinispanClientProcessor.java
Comment (karesti) [reply extensions/infinispan-client/deployment/src/main/java/io/quarkus/infinispan/client/deployment/InfinispanClientProcessor.java:245]: @wburns @gwenneg Indeed I don't know if there are cases where we will face issues if I don't put those here. My tests work with the 3 annotations and I forgot one, so I don't know if we need that here...

File: extensions/infinispan-client/deployment/src/main/java/io/quarkus/infinispan/client/deployment/cache/CacheDeploymentConstants.java
Comment (karesti) [reply extensions/infinispan-client/deployment/src/main/java/io/quarkus/infinispan/client/deployment/cache/CacheDeploymentConstants.java:14]: yes, I did not do any validation on the build time part so the constants are not needed for now.

File: extensions/infinispan-client/deployment/src/main/java/io/quarkus/infinispan/client/deployment/cache/exception/ClassTargetException.java
Comment (karesti) [reply extensions/infinispan-client/deployment/src/main/java/io/quarkus/infinispan/client/deployment/cache/exception/ClassTargetException.java:16]: same as before. I skipped the validation part in deployment and improve that in a second PR with null values support as well

File: extensions/infinispan-client/runtime/src/main/java/io/quarkus/infinispan/client/CacheResult.java
Comment (karesti) [reply extensions/infinispan-client/runtime/src/main/java/io/quarkus/infinispan/client/CacheResult.java:14]: ok

File: extensions/infinispan-client/runtime/src/main/java/io/quarkus/infinispan/client/runtime/cache/CacheInterceptor.java
Comment (karesti) [reply extensions/infinispan-client/runtime/src/main/java/io/quarkus/infinispan/client/runtime/cache/CacheInterceptor.java:22]: ok!

**Comment by karesti** [comment]
@gwenneg I made locally the small review changes and I will check the 2 PR you mentioned to report the changes here too. I ping you once is ready

**Comment by karesti** [comment]
@geoand I've been sidetracked too much recently, I need to report the changes reported by @gwenneg

**Comment by karesti** [comment]
@geoand give me some days to get back to it properly. Sorry again

**Comment by geoand** [comment]
NP!

**Comment by karesti** [comment]
@geoand @gwenneg FYI coming back to work on this fully for the next days (and other features preparing the integration with Infinispan 14)

**Comment by geoand** [comment]
👍

**Comment by karesti** [comment]
@geoand I need to check the failing test :) I did not run them properly as I thought yesterday, sorry about it

**Comment by geoand** [comment]
No problem :)

**Comment by geoand** [comment]
CI is passing. Is this ready for a review?

**Comment by geoand** [comment]
One thing I would like us to have in mind for a follow-up, is to provider proper metrics

**Comment by karesti** [comment]
@geoand yes please

**Comment by karesti** [comment]
> One thing I would like us to have in mind for a follow-up, is to provider proper metrics (as we do for Caffeine)

yes, we can totally create an issue to make it happen after that

File: extensions/infinispan-client/deployment/src/main/java/io/quarkus/infinispan/client/deployment/InfinispanClientProcessor.java
Comment (gwenneg) [reply extensions/infinispan-client/deployment/src/main/java/io/quarkus/infinispan/client/deployment/InfinispanClientProcessor.java:245]: The interceptors are already declared as unremovable above so I'd say that they don't need to be added here. I'm not sure though. :smiley:

File: extensions/infinispan-client/runtime/src/main/java/io/quarkus/infinispan/client/runtime/cache/CacheInvalidateAllInterceptor.java
Comment (gwenneg) [inline extensions/infinispan-client/runtime/src/main/java/io/quarkus/infinispan/client/runtime/cache/CacheInvalidateAllInterceptor.java:57]: Some details from #24894 are missing in this class (at this line but also at other lines in the class).

File: extensions/infinispan-client/runtime/src/main/java/io/quarkus/infinispan/client/runtime/cache/CacheInvalidateInterceptor.java
Comment (gwenneg) [inline extensions/infinispan-client/runtime/src/main/java/io/quarkus/infinispan/client/runtime/cache/CacheInvalidateInterceptor.java:59]: Same comment about #24894.

File: extensions/infinispan-client/runtime/src/main/java/io/quarkus/infinispan/client/runtime/cache/CacheResultInterceptor.java
Comment (gwenneg) [inline extensions/infinispan-client/runtime/src/main/java/io/quarkus/infinispan/client/runtime/cache/CacheResultInterceptor.java:95]: Same comment about #24894.

**Comment by geoand** [comment]
Dump question that has probably been answered before: Do we really want to have a separate set of Infinispan specific annotations instead of reusing the Quarkus Cache annotations?

**Comment by gwenneg** [comment]
Assuming that's possible (I didn't check), Infinispan could even be a caching provider from quarkus-cache that could be enabled from the quarkus-cache config when the Infinispan extension is on the classpath. Quarkus-cache was built with that kind of integration case in mind.

Unfortunately, this PR already shows that maintaining the duplicated code (annotations and interceptors) won't be an easy task.

**Comment by karesti** [comment]
@gwenneg there are limitations concerning the local cache of caffeine and how it works with infinispan. 
For now, I think we need to implement this feature appart from quarkus-cache and then try to implement caching providers of course. 

I will report the changes @geoand and @gwenneg hopefully this will be enough to integrate the feature in the infinispan extension

**Comment by geoand** [comment]
Then maybe we need a set of annotation for remote caching specifically?

**Comment by karesti** [comment]
@geoand we might want to think about this but for now the Infinispan team would really like to offer this first implementation in the extension and improve the use through users as well. This feature is taking way too long to overthink it too much right now IMO

**Comment by geoand** [comment]
Yeah, I understand, but would it not make sense to provide the remote caching annotations now? 
It should not be too much work (theoretically) and it would shield us from moving / breaking the annotations in the future.

I won't block the PR on this suggestion, but I highly suggest it should be given serious thought (cc @gsmet)

File: extensions/infinispan-client/runtime/src/main/java/io/quarkus/infinispan/client/runtime/cache/CacheInvalidateAllInterceptor.java
Comment (gwenneg) [reply extensions/infinispan-client/runtime/src/main/java/io/quarkus/infinispan/client/runtime/cache/CacheInvalidateAllInterceptor.java:57]: This comment and two other ones were marked as resolved but I don't see any changes. Is the fix about to be committed?

**Comment by karesti** [comment]
@geoand I agree on working together towards that goal. I'm reporting the missing changes now, I ping you once it's done!
cc @gwenneg

**Comment by geoand** [comment]
Thanks!

File: extensions/infinispan-client/runtime/src/main/java/io/quarkus/infinispan/client/runtime/cache/CacheInvalidateAllInterceptor.java
Comment (karesti) [reply extensions/infinispan-client/runtime/src/main/java/io/quarkus/infinispan/client/runtime/cache/CacheInvalidateAllInterceptor.java:57]: yes, I have ot locally ;) is for me to mark it as done

**Comment by karesti** [comment]
@geoand @gwenneg how does it look now ?

**Comment by geoand** [comment]
I'll be on PTO for the next 10 days or so

**Comment by gwenneg** [comment]
I just came back from PTO. I'll have a look at this PR later today.

File: extensions/infinispan-client/runtime/src/main/java/io/quarkus/infinispan/client/runtime/cache/InfinispanGetWrapper.java
Comment (gwenneg) [inline extensions/infinispan-client/runtime/src/main/java/io/quarkus/infinispan/client/runtime/cache/InfinispanGetWrapper.java:44]: Why is the lock logic needed here? A new `InfinispanGetWrapper` instance is created on each `CacheResultInterceptor#intercept` invocation so I can't see when this lock will actually prevent a cache value from being stored in the remote cache using the value loader.

The lock logic would make more sense to me if the locks were stored in a "persistent" (from a memory perspective) wrapper instance around `RemoteCache`, but maybe I missed something.

**Review by gwenneg — COMMENTED** [review]
All changes from #24894 are here now :+1:

File: extensions/infinispan-client/runtime/src/main/java/io/quarkus/infinispan/client/runtime/cache/InfinispanGetWrapper.java
Comment (karesti) [reply extensions/infinispan-client/runtime/src/main/java/io/quarkus/infinispan/client/runtime/cache/InfinispanGetWrapper.java:44]: You are right on this, since the object is created each time.
This is me extracting the code and then not realizing on this part haha thanks for that
we need the lock from the client perspective as we do in the infinispan spring-boot integration.
let me fix this and ping you back 
thanks!

File: extensions/infinispan-client/runtime/src/main/java/io/quarkus/infinispan/client/runtime/cache/CacheInvalidateAllInterceptor.java
Comment (karesti) [reply extensions/infinispan-client/runtime/src/main/java/io/quarkus/infinispan/client/runtime/cache/CacheInvalidateAllInterceptor.java:78]: thanks

**Comment by karesti** [comment]
@geoand this should be ready now

File: extensions/infinispan-client/runtime/src/main/java/io/quarkus/infinispan/client/runtime/cache/CacheResultInterceptor.java
Comment (gwenneg) [inline extensions/infinispan-client/runtime/src/main/java/io/quarkus/infinispan/client/runtime/cache/CacheResultInterceptor.java:31]: @mkouba Will that `Map` be shared across all interceptions?

Comment (gwenneg) [reply extensions/infinispan-client/runtime/src/main/java/io/quarkus/infinispan/client/runtime/cache/CacheResultInterceptor.java:31]: @karesti Isn't that `Map` supposed to lock cache keys? What if two different caches are invoked at the same time with the same key? One will be locked waiting for the other while it's not needed, right?

Comment (mkouba) [reply extensions/infinispan-client/runtime/src/main/java/io/quarkus/infinispan/client/runtime/cache/CacheResultInterceptor.java:31]: > @mkouba Will that `Map` be shared across all interceptions of that interceptor?

Not really. This map will be used for all interceptions of a _particular bean instance_. 

Instances of interceptors are _dependent objects_ of the bean instance they intercept. In other words, whenever a bean instance is created a relevant interceptor instance is created as well and bound to the lifecycle of the intercepted bean.

File: docs/src/main/asciidoc/infinispan-client.adoc
Comment (geoand) [inline docs/src/main/asciidoc/infinispan-client.adoc:445]: I don't see the fully qualified name anywhere in this section. I am pretty sure that will lead users to think that they can use the regular Quarkus caching annotations

File: extensions/infinispan-client/runtime/src/main/java/io/quarkus/infinispan/client/runtime/cache/CacheResultInterceptor.java
Comment (karesti) [reply extensions/infinispan-client/runtime/src/main/java/io/quarkus/infinispan/client/runtime/cache/CacheResultInterceptor.java]: So if I have 2 application scope endpoints I have 2 different interceptors if I use CacheResult annotation. Is that allright ? @mkouba

File: extensions/infinispan-client/runtime/src/main/java/io/quarkus/infinispan/client/runtime/cache/InfinispanGetWrapper.java
Comment (wburns) [inline extensions/infinispan-client/runtime/src/main/java/io/quarkus/infinispan/client/runtime/cache/InfinispanGetWrapper.java:17]: This whole class needs to be rewritten to be non blocking. All of the Cache operations in `getFromInfinispan` are blocking.
I would suggest changing this to be `Map<Object, CompletionStage<Object>>` to allow for that.

File: extensions/infinispan-client/runtime/src/main/java/io/quarkus/infinispan/client/runtime/cache/InfinispanGetWrapper.java
Comment (wburns) [inline extensions/infinispan-client/runtime/src/main/java/io/quarkus/infinispan/client/runtime/cache/InfinispanGetWrapper.java:43]: This code is blocking. We need to do something like instead

To do coalescing of operations I would suggest doing the following

```java
CompletionStage<Object> stage = new CompletionStage<>();
CompletionStage<Object> prev;
if ((prev = synchronousGetLocks.putIfAbsent(key, stage)) != null) {
   return prev;
}
cache.getAsync().whenComplete((v, t) -> {
   if (t != null) {
      stage.completeExceptionally(t);
   } else if (v != null) {
      stage.complete(v);
   } else {
      Object calculatedValue = ...
      if (calculatedValue != null) {
         cache.putIfAbsent(key, calculatedValue)
            .whenComplete((v, t) -> {
                if (t != null) {
                  stage.completeExceptionally(t);
                } else {
                  stage.complete(v == null ? calculatedValue : v);
                }
                synchronousGetLocks.remove(key);
            });
         return;
      }
   }
   synchronousGetLocks.remove(key);
});
return stage;

```

Comment (wburns) [reply extensions/infinispan-client/runtime/src/main/java/io/quarkus/infinispan/client/runtime/cache/InfinispanGetWrapper.java:43]: To be 100% safe you may want to add a try/catch block around the entire thing and remove from `synchronousGetLocks` if any Throwable is thrown.

**Review by wburns — CHANGES_REQUESTED** [review]

File: extensions/infinispan-client/runtime/src/main/java/io/quarkus/infinispan/client/runtime/cache/InfinispanGetWrapper.java
Comment (karesti) [reply extensions/infinispan-client/runtime/src/main/java/io/quarkus/infinispan/client/runtime/cache/InfinispanGetWrapper.java:30]: I got inspired from the CacheExtension for this part that does not use lambda

Comment (karesti) [reply extensions/infinispan-client/runtime/src/main/java/io/quarkus/infinispan/client/runtime/cache/InfinispanGetWrapper.java:30]: I will move it to lambda

Comment (gwenneg) [reply extensions/infinispan-client/runtime/src/main/java/io/quarkus/infinispan/client/runtime/cache/InfinispanGetWrapper.java:30]: Last time I checked lambdas were kind of "forbidden" in all Quarkus runtime code (but allowed in the build time code).

File: extensions/infinispan-client/runtime/src/main/java/io/quarkus/infinispan/client/runtime/cache/CacheResultInterceptor.java
Comment (karesti) [reply extensions/infinispan-client/runtime/src/main/java/io/quarkus/infinispan/client/runtime/cache/CacheResultInterceptor.java:142]: not necessary

File: extensions/infinispan-client/runtime/src/main/java/io/quarkus/infinispan/client/runtime/cache/CacheResultInterceptor.java
Comment (wburns) [reply extensions/infinispan-client/runtime/src/main/java/io/quarkus/infinispan/client/runtime/cache/CacheResultInterceptor.java]: Yeah, you would want it isolated per bean since each bean is tied to a specific cache. Otherwise keys from different caches could return the others values.

**Comment by karesti** [comment]
@wburns ready for a review round

File: extensions/infinispan-client/runtime/src/main/java/io/quarkus/infinispan/client/runtime/cache/InfinispanGetWrapper.java
Comment (wburns) [inline extensions/infinispan-client/runtime/src/main/java/io/quarkus/infinispan/client/runtime/cache/InfinispanGetWrapper.java:53]: Should we throw an exception if the value is null? I kinda wonder if this cache annotation requires non null values to prevent misses for the same key causing remote lookups and recalculations

File: extensions/infinispan-client/runtime/src/main/java/io/quarkus/infinispan/client/runtime/cache/InfinispanGetWrapper.java
Comment (wburns) [inline extensions/infinispan-client/runtime/src/main/java/io/quarkus/infinispan/client/runtime/cache/InfinispanGetWrapper.java:64]: This should be removed when the stage is completed. We are removing it too early here imo.

Thinking about this more, should we just confirm if we want to wait until the write completes before returning the value or return early. The latter will reduce latency for an individual request, but if you have concurrent requests it would be possible, but unlikely, for interleaving to do the calculation twice.

**Review by wburns — COMMENTED** [review]
Just a couple more comments, otherwise looks good to me.

File: extensions/infinispan-client/runtime/src/main/java/io/quarkus/infinispan/client/runtime/cache/InfinispanGetWrapper.java
Comment (karesti) [reply extensions/infinispan-client/runtime/src/main/java/io/quarkus/infinispan/client/runtime/cache/InfinispanGetWrapper.java:53]: So, I think this comes into the debate of "supporting null values" caching or not. Spring Cache supports null values caching (when enabled). In this case we are not *yet* supporting it. In my opinion nothing should be done at this point

File: extensions/infinispan-client/runtime/src/main/java/io/quarkus/infinispan/client/runtime/cache/InfinispanGetWrapper.java
Comment (wburns) [reply extensions/infinispan-client/runtime/src/main/java/io/quarkus/infinispan/client/runtime/cache/InfinispanGetWrapper.java:29]: @gwenneg Do you have more info on this? I would to see the discussion regarding it. As far as I know the memory overhead was mostly fixed where it doesn't require another class I thought.

File: extensions/infinispan-client/runtime/src/main/java/io/quarkus/infinispan/client/runtime/cache/InfinispanGetWrapper.java
Comment (wburns) [inline extensions/infinispan-client/runtime/src/main/java/io/quarkus/infinispan/client/runtime/cache/InfinispanGetWrapper.java:54]: Just noticed, this is missing a `stage.complete(null)`. Guessing we don't have any coverage for concurrent misses on the cache.

**Comment by karesti** [comment]
@wburns changes reported 
@geoand

**Review by wburns — APPROVED** [review]
LGTM from the ISPN side

**Comment by geoand** [comment]
Test failures seem related

File: extensions/infinispan-client/runtime/src/main/java/io/quarkus/infinispan/client/runtime/cache/CacheResultInterceptor.java
Comment (gwenneg) [inline extensions/infinispan-client/runtime/src/main/java/io/quarkus/infinispan/client/runtime/cache/CacheResultInterceptor.java:31]: I'd like to confirm something based on this:
```
@ApplicationScoped
class A {
    @CacheResult(cacheName = "foo")
    String getValue(Object key) { /** body omitted */ }
}

@Singleton
class B {
    @CacheResult(cacheName = "foo")
    String getValue(Object key) { /** body omitted */ }
}
```

Unless I misunderstood what @mkouba wrote, if `A#getValue` and `B#getValue` are called concurrently with the same arg, the remote Infinispan cache will _not_ be protected against concurrent read/write access because each bean (A or B) will have its own instance of `CacheResultInterceptor` and therefore its own locks Map. Doesn't that defeat the very reason this Map exists? Or maybe I missed something.

Comment (karesti) [reply extensions/infinispan-client/runtime/src/main/java/io/quarkus/infinispan/client/runtime/cache/CacheResultInterceptor.java:31]: Unless an easy solution exists, probably we should consider writing down that limitation/risk and write down some  of good practices. WDYT?

Comment (gwenneg) [reply extensions/infinispan-client/runtime/src/main/java/io/quarkus/infinispan/client/runtime/cache/CacheResultInterceptor.java:31]: The Map could probably be moved to a separate `@ApplicationScoped` or `@Singleton` bean and that bean could then be injected in the interceptors.

Comment (karesti) [reply extensions/infinispan-client/runtime/src/main/java/io/quarkus/infinispan/client/runtime/cache/CacheResultInterceptor.java:31]: obviously @gwenneg 🤦‍♀️ ;) thanks will do that

Comment (karesti) [reply extensions/infinispan-client/runtime/src/main/java/io/quarkus/infinispan/client/runtime/cache/CacheResultInterceptor.java]: I added a change to scope the map into a bean @gwenneg 
let me know if I missed something please

File: extensions/infinispan-client/runtime/src/main/java/io/quarkus/infinispan/client/runtime/cache/SynchronousInfinispanGet.java
Comment (gwenneg) [inline extensions/infinispan-client/runtime/src/main/java/io/quarkus/infinispan/client/runtime/cache/SynchronousInfinispanGet.java:19]: This could be simplified a bit:
```suggestion
    public Map<Object, CompletableFuture<Object>> get(String cacheName) {
        return synchronousGetLocks.computeIfAbsent(cacheName, k -> new ConcurrentHashMap<>());
    }
```

**Review by gwenneg — APPROVED** [review]
I added a minor comment below.

LGTM

File: extensions/infinispan-client/runtime/src/main/java/io/quarkus/infinispan/client/runtime/cache/SynchronousInfinispanGet.java
Comment (karesti) [reply extensions/infinispan-client/runtime/src/main/java/io/quarkus/infinispan/client/runtime/cache/SynchronousInfinispanGet.java]: thanks!

**Comment by karesti** [comment]
@geoand this seems validated

**Comment by geoand** [comment]
Cool! I'll let @gwenneg merge this

**Pull request merged**
